features/netsocket/cellular/generic_modem_driver/TESTS/unit_tests/default/main.cpp

```diff
@@ -0,0 +1,449 @@
+            success = true;
+        }
+    }
+    TEST_ASSERT(success);
+
+    free(recv_data);
+}
+
+/**
+ * Send an entire TCP data buffer until done
+ */
+static int send_all(TCPSocket *sock, const char *data, int size)
+{
+    int x;
+    int count = 0;
+    Timer timer;
+
+    timer.start();
+    while ((count < size) && (timer.read_ms() < ASYNC_TEST_WAIT_TIME)) {
+        x = sock->send(data + count, size - count);
+        if (x > 0) {
+            count += x;
+            tr_debug("%d byte(s) sent, %d left to send.", count, size - count);
+        }
+        wait_ms(10);
+    }
+    timer.stop();
+
+    return count;
+}
+
+/**
+ * The asynchronous callback
+ */
+static void async_cb(bool *callback_triggered)
+{
+
+    TEST_ASSERT(callback_triggered != NULL);
+    *callback_triggered = true;
+}
+
+/**
+ * Do a TCP echo using the asynchronous driver
+ */
+static void do_tcp_echo_async(TCPSocket *sock, int size,
+                              bool *callback_triggered)
+{
+    void * recv_data = malloc(size);
+    int recv_size = 0;
+    int remaining_size;
+    int x, y;
+    Timer timer;
+    TEST_ASSERT(recv_data != NULL);
+
+    *callback_triggered = false;
+    tr_debug("Echo testing TCP packet size %d byte(s) async.", size);
+    TEST_ASSERT(send_all(sock, test_data, size) == size);
+    // Wait for all the echoed data to arrive
+    timer.start();
+    remaining_size = size;
+    while ((recv_size < size) && (timer.read_ms() < ASYNC_TEST_WAIT_TIME)) {
+        if (*callback_triggered) {
+            *callback_triggered = false;
+            x = sock->recv((char *) recv_data + recv_size, remaining_size);
+            if (x > 0) {
+                recv_size += x;
+                remaining_size = size - recv_size;
+                tr_debug("%d byte(s) echoed back so far, %d to go.", recv_size,
+                         remaining_size);
+            }
+        }
+        wait_ms(10);
+    }
+    TEST_ASSERT(recv_size == size);
+    y = memcmp(test_data, recv_data, size);
+    if (y != 0) {
+        tr_debug("Sent %d, |%*.*s|", size, size, size, test_data);
+        tr_debug("Rcvd %d, |%*.*s|", size, size, size, (char * ) recv_data);
+        // We do not assert a failure here because ublox TCP echo server doesn't send
+        // back original data. It actually constructs a ublox message string. They need to fix it as
+        // at the minute in case of TCP, their server is not behaving like a echo TCP server.
+        //TEST_ASSERT(false);
+    }
+    timer.stop();
+    free(recv_data);
+}
+
+/**
+ * Use a connection, checking that it is good
+ * Checks via doing an NTP transaction
+ */
+static void use_connection(OnboardCellularInterface *driver)
+{
+    const char * ip_address = driver->get_ip_address();
+    const char * net_mask = driver->get_netmask();
+    const char * gateway = driver->get_gateway();
+
+    TEST_ASSERT(driver->is_connected());
+
+    TEST_ASSERT(ip_address != NULL);
+    tr_debug("IP address %s.", ip_address);
```

Comment:
Yes, its a connectivity test. If you are connected, you should have an IP address as well.